pkg/util/wsstream/conn.go

```diff
@@ -0,0 +1,306 @@
+func IsWebSocketRequest(req *http.Request) bool {
+	return connectionUpgradeRegex.MatchString(strings.ToLower(req.Header.Get("Connection"))) && strings.ToLower(req.Header.Get("Upgrade")) == "websocket"
+}
+
+// ignoreReceives reads from a WebSocket until it is closed, then returns. If timeout is set, the
+// read and write deadlines are pushed every time a new message is received.
+func ignoreReceives(ws *websocket.Conn, timeout time.Duration) {
+	defer util.HandleCrash()
+	var data []byte
+	for {
+		resetTimeout(ws, timeout)
+		if err := websocket.Message.Receive(ws, &data); err != nil {
+			if err == io.EOF {
+				return
+			}
+		}
+	}
+}
+
+// handshake ensures the provided user protocol matches one of the allowed protocols. It returns
+// no error if no protocol is specified.
+func handshake(config *websocket.Config, req *http.Request, allowed []string) error {
+	protocols := config.Protocol
+	if len(protocols) == 0 {
+		return nil
+	}
+	for _, protocol := range protocols {
+		for _, allow := range allowed {
+			if allow == protocol {
+				config.Protocol = []string{protocol}
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("requested protocol(s) are not supported: %v; supports %v", config.Protocol, allowed)
+}
+
+// Conn supports sending multiple binary channels over a websocket connection.
+// Supports only the "channel.k8s.io" subprotocol.
+type Conn struct {
+	channels []*websocketChannel
+	codec    codecType
+	ready    chan struct{}
+	ws       *websocket.Conn
+	timeout  time.Duration
+}
+
+// NewConn creates a WebSocket connection that supports a set of channels. Channels begin each
+// web socket message with a single byte indicating the channel number (0-N). 255 is reserved for
+// future use. The channel types for each channel are passed as an array, supporting the different
+// duplex modes. Read and Write refer to whether the channel can be used as a Reader or Writer.
+func NewConn(channels ...ChannelType) *Conn {
+	conn := &Conn{
+		ready:    make(chan struct{}),
+		channels: make([]*websocketChannel, len(channels)),
+	}
+	for i := range conn.channels {
+		switch channels[i] {
+		case ReadChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), true, false)
+		case WriteChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), false, true)
+		case ReadWriteChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), true, true)
+		case IgnoreChannel:
+			conn.channels[i] = newWebsocketChannel(conn, byte(i), false, false)
+		}
+	}
+	return conn
+}
+
+// SetIdleTimeout sets the interval for both reads and writes before timeout. If not specified,
+// there is no timeout on the connection.
+func (conn *Conn) SetIdleTimeout(duration time.Duration) {
+	conn.timeout = duration
+}
+
+// Open the connection and create channels for reading and writing.
+func (conn *Conn) Open(w http.ResponseWriter, req *http.Request) ([]io.ReadWriteCloser, error) {
+	go func() {
+		defer util.HandleCrash()
+		defer conn.Close()
```

Comment:
If the websocket server (not our code) exits without invoking our code, the connection has to be closed.  Test code directly invokes handle.  The call is reentrant, and both places guarantee we properly close the WS.